File: q2/code.go
package q2

type ListNode struct {
    Val int
    Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var ret *ListNode
	var prev *ListNode

	next := 0

	for ;; {
		var current *ListNode = nil
		sum := 0

		// increment & fetch current value
		if l1 != nil {
			// reuse
			current = l1
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			// reuse
			current = l2
			sum += l2.Val
			l2 = l2.Next
		}

		// check exit condition
		if current == nil {
			break
		}

		// calc sum
		sum += next
		if sum > 9 {
			next = 1
			current.Val = sum - 10
		} else {
			next = 0
			current.Val = sum
		}

		// set current pointer
		if prev != nil {
			prev.Next = current
		} else {
			ret = current
		}
		prev = current
	}

	if next == 1 {
		prev.Next = new(ListNode)
		prev.Next.Val = 1
	}

	return ret;
}

File: q2/code.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var carry = 0, ret = [];

    for (;;) {
        var isExit = true;
        var sum = 0;
        if (l1) {
            isExit = false;
            sum += l1.val|0;
            l1 = l1.next;
        }
        if (l2) {
            isExit = false;
            sum += l2.val|0;
            l2 = l2.next;
        }
        if (isExit) break;

        sum += carry;
        if (sum > 9) {
            carry = 1;
            ret[ret.length] = sum - 10;
        } else {
            carry = 0;
            ret[ret.length] = sum;
        }
    }

    if (carry) { ret[ret.length] = 1; }

    return ret;
};
